Add ability to disconnect selected Bluetooth devices

ramatah opened the pull request:
These changes add the ability to disconnect selected Bluetooth devices instead of powering down the Bluetooth stack. It can also restore their states on screen unlock.

I'm using a Mac mini and I needed a way to disconnect my Bluetooth headphones on sleep. But I still needed to be able to wake up my Mac with my Bluetooth mouse and my Bluetooth keyboard.

I'm not a Mac developer so I have no pretension that what I implemented is correct or not. I just needed a quick solution to my problem.

![screenshot](https://user-images.githubusercontent.com/41805917/222552018-f6db4cc6-d869-42a2-b257-a294f44c2071.png)

ramatah committed:
Add ability to disconnect Bluetooth devices

- Instead of fully powering down the Bluetooth stack on sleep, allow to select a list of devices that should be disconnected instead.
- The state of the disconnected device can be restored on screen unlock.
- Only Bluetooth Classic devices are supported -- no BLE devices.
  IOBluetoothDevice.pairedDevices() has inconsistent behaviors.

ramatah committed:
Update screenshot to include "disconnect Bluetooth devices" feature.

ramatah committed:
Update README to include the "disconnect Bluetooth devices" settings.

ramatah committed:
Code cleanup

ramatah committed:
Use os_log() to log messages instead of print().

ramatah committed:
Fix connecting to Bluetooth after screen unlock when "action == enable".

stefansundin commented:
Wow, this looks great. I'll merge this and release a new version after I've had a chance to try it out.

stefansundin commented:
> I'm not a Mac developer so I have no pretension that what I implemented is correct or not. I just needed a quick solution to my problem.

Neither am I. 😆

ramatah commented:
😂 I encountered a weird behaviour with getting the list of paired devices -- which I documented in the `getPairedDevices()` function.

ramatah committed:
- Fix wrong log when logging wifi related message.
- Defer the string conversion of the argument when using os_log instead of eagerly calling .description.

ramatah committed:
Improve logging when setting Wifi status

ramatah committed:
Fix race condition where Bluetooth devices can remain disconnected after quickly waking up from sleep.

ramatah committed:
Code cleanup

stefansundin commented:
It looks like you are still finding some new changes that you want to make so I won't merge this just yet. But I did check out your code and am running this version now, so I will help you test. We can plan on merging it in 2-3 weeks if things are working well. I am not in a rush to get it merged, better to have a stable release.

One thing I noticed right away is that I think it would be great if you could click `Disconnect selected devices` when in this state:

![Screenshot 2023-03-04 at 15 46 50](https://user-images.githubusercontent.com/1991151/222934194-d18b8561-7331-4dff-80d7-ddbc5d68143c.png)

It is possible to go back to `Turn off Bluetooth` so we should be able to switch both ways easily.

![Screenshot 2023-03-04 at 15 51 17](https://user-images.githubusercontent.com/1991151/222934218-edc5c8d5-5390-4e99-a15f-37f50d309aed.png)

Will let you know if I see anything else.

stefansundin commented:
I pushed two small commits to bring some of your changes into the main branch. Feel free to rebase.

stefansundin commented:
A while ago I wrote some logging code to try to figure out #5 (it hasn't helped me figure out the issue yet though). I just pushed it to a branch now for posterity: https://github.com/stefansundin/bluesnooze/commits/logging

I wonder how `os.log` compares. I assume that macOS persists these logs somewhere? I looked but I wasn't able to find it. It would be great if it stored it e.g. in `~/Library/Logs/` so that users could submit a log if the program doesn't work as expected for them.

I'm considering bringing over the `setWifi` changes to my branch first but would be great if we could decide on the logging first. The only good thing about what I wrote is that it puts it in a file, but otherwise it is very clunky (I was not planning on including it in a release). If we find something that can be included in a release that would be great.

If we bring over the less scary stuff first then we can reduce this PR to only have the new Bluetooth feature.

ramatah committed:
Merge remote-tracking branch 'stefansundin/main' into main

ramatah committed:
Allow to select "Bluetooth > Disconnect selected devices" without having to deselect "Turn off devices" first.

ramatah committed:
Expose the retry interval of waitToBeActuallyDisconnected() as a parameter.

ramatah commented:
> One thing I noticed right away is that I think it would be great if you could click `Disconnect selected devices` when in this state:

I agree. I pushed a change that implements that behaviour.

ramatah commented:
> I wonder how `os.log` compares. I assume that macOS persists these logs somewhere?

Indeed, it's the macOS unified logging system and it takes care of storing the logs. You can view them with the `Console` app or with the `log` command line tool. When running the app in Xcode, the logs will be streamed directly in the debug output.

As per the `os_log` man page:
> The unified logging system provides a single, efficient, high performance set of APIs for capturing log messages across all levels of the system.  This unified system centralizes the storage of log data in memory and in a data store on disk.  The system implements global settings that govern logging behavior and persistence, while at the same time providing fine-grained control during debugging via the log(1) command-line tool and through the use of custom logging configuration profiles (see os_log(5)). Log messages are viewed using the Console app in /Applications/Utilities/ and the log(1) command-line tool.

From my experience on macOS Ventura 13.2.1:
1. The `Console` app allows to live stream the logs.  
    It has some nice filtering capabilities to help in viewing the logs.
3. But I was confused as to how you can see the logs that weren't live streamed.
4. This is where the `log` command comes to play.
    1. `log show --process Bluesnooze` will show all the logs of the Bluesnooze process.
    2. `sudo log collect` can export all the system logs into an archive that can then be opened in the `Console` app.
5. `log` has options to control the time window of the logs that are of interest.

What's pretty cool about it is that it will not only include the logs from our code, but it will also include the logs of the frameworks that are used by the process. For example, it will include the logs from the `IOBluetooth` framework.

> but would be great if we could decide on the logging first.

I suggest using the `os_log` as it's the standard logging system.

**References:**
- os_log(3), https://www.manpagez.com/man/3/os_log/
- log(1), https://www.manpagez.com/man/1/log/

ramatah commented:
> A while ago I wrote some logging code to try to figure out https://github.com/stefansundin/bluesnooze/issues/5 (it hasn't helped me figure out the issue yet though).

With `log collect`, you can dump the system logs into a log archive that you can then open in the `Console` app. It will allow you to see the logs of the other processes around the time the app tries to turn the Wifi back on.

For example, here are some sample logs that show the interactions between the app and the `bluetoothd` when the app tries to connect to a Bluetooth device:
```
default	2023-03-13 12:17:11.373534 -0400	Bluesnooze	Connect <IOBluetoothDevice: 0x121804bd0 ATH-M50xBT2, 00-0a-45-23-cd-19>, paired: true, connected: false
default	2023-03-13 12:17:11.373579 -0400	bluetoothd	Received XPC message "CBMsgIdRetrievePairedPeersWithOptions" from session "com.oliverpeate.Bluesnooze-classic-72022-832"
default	2023-03-13 12:17:11.375304 -0400	bluetoothd	Received XPC message "CBMsgIdPairingAgentRetrievePairedDevices" from session "com.oliverpeate.Bluesnooze-central-72022-833"
default	2023-03-13 12:17:11.375570 -0400	Bluesnooze	-[IOBluetoothDevice openConnection:withPageTimeout:authenticationRequired:allowRoleSwitch:forPairing:] async: <private>
default	2023-03-13 12:17:11.375576 -0400	Bluesnooze	-[IOBluetoothCoreBluetoothCoordinator connectPeer:options:] <private>
default	2023-03-13 12:17:11.375588 -0400	Bluesnooze	-[IOBluetoothDevice openConnection:withPageTimeout:authenticationRequired:allowRoleSwitch:forPairing:] Success: <private>
default	2023-03-13 12:17:11.375622 -0400	Bluesnooze	<IOBluetoothDevice: 0x121804bd0 ATH-M50xBT2, 00-0a-45-23-cd-19>, CREATE_CONNECTION command success: true
default	2023-03-13 12:17:11.376737 -0400	Bluesnooze	-[CBClassicManager connectPeer:options:] <private> uuids: <private>
default	2023-03-13 12:17:11.378319 -0400	bluetoothd	Received XPC message "CBMsgIdConnectPeer" from session "com.oliverpeate.Bluesnooze-classic-72022-832"
default	2023-03-13 12:17:11.391827 -0400	bluetoothd	UN: connectDeviceForApp for classic  8EFEC9E2-0BB9-823C-A585-96101514383F for com.oliverpeate.Bluesnooze
default	2023-03-13 12:17:11.391838 -0400	bluetoothd	App connecting - {cbuuid: 8EFEC9E2-0BB9-823C-A585-96101514383F, bundle: com.oliverpeate.Bluesnooze}
default	2023-03-13 12:17:11.391846 -0400	bluetoothd	Connecting device "8EFEC9E2-0BB9-823C-A585-96101514383F" for session "com.oliverpeate.Bluesnooze-classic-72022-832" with options (connect:0 disconnect:0 notify:0 delay:0 wantsObjectDiscoverData:0 transportBridging:0 disableLeGATT:0 ancsRequired:0)
[...]
default	2023-03-13 12:17:13.886193 -0400	bluetoothd	Sending XPC message "CBMsgIdPeerConnectionCompleted" to session "com.oliverpeate.Bluesnooze-classic-72022-832"
default	2023-03-13 12:17:13.886611 -0400	Bluesnooze	-[IOBluetoothCoreBluetoothCoordinator init]_block_invoke connectionStatusHandler callback for peer: <private> with error: 0
default	2023-03-13 12:17:13.886616 -0400	Bluesnooze	-[IOBluetoothCoreBluetoothCoordinator init]_block_invoke connectionStatusHandler: <private> with error: 0
default	2023-03-13 12:17:13.886674 -0400	Bluesnooze	-[IOBluetoothDevice peerConnected:error:] peer: <private> error:0
default	2023-03-13 12:17:13.886743 -0400	Bluesnooze	-[CBClassicManager handlePeerConnectionCompleted:] <private> <private>
default	2023-03-13 12:17:13.886749 -0400	Bluesnooze	Inform delegate connectionComplete: 0 <private> <<private>>
default	2023-03-13 12:17:13.886821 -0400	Bluesnooze	Connection complete: <IOBluetoothDevice: 0x121804bd0 ATH-M50xBT2, 00-0a-45-23-cd-19>, success: true
```

**Notes:**
- The documentation doesn't mention it but `log collect` requires `root` so you need to `sudo log collect`.
- When you open the log archive in the `Console` app, by default, it will only show the last 5 minutes.
   This can be very confusing, but there's a dropdown when you can change the time window.

Faesh commented:
Hey, I was wondering if it is possible to download compiled file of this version.

Atemu commented:
Hi, what's the status on this PR?
